### PR TITLE
kvserver: resurrect using_applied_state_key in ReplicaState

### LINKED ISSUE #72029: roachtest: version/mixed/nodes=5 failed
roachtest.version/mixed/nodes=5 [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3641256&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3641256&tab=artifacts#/version/mixed/nodes=5) on master @ [965b99b1a05eb893ee908d852680824a0bd4ffe3](https://github.com/cockroachdb/cockroach/commits/965b99b1a05eb893ee908d852680824a0bd4ffe3):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/version/mixed/nodes=5/run_1
	monitor.go:128,version.go:211,version.go:224,test_runner.go:777: monitor failure: unexpected node event: 3: dead (exit status 7)
		(1) attached stack trace
		  -- stack trace:
		  | main.(*monitorImpl).WaitE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:116
		  | main.(*monitorImpl).Wait
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:124
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerVersion.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/version.go:211
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerVersion.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/version.go:224
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:777
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (2) monitor failure
		Wraps: (3) unexpected node event: 3: dead (exit status 7)
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errors.errorString

	cluster.go:1300,context.go:91,cluster.go:1288,test_runner.go:866: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-3641256-1635315289-38-n6cpu4 --oneshot --ignore-empty-nodes: exit status 1 3: dead (exit status 7)
		6: skipped
		5: 17957
		4: 18653
		2: 20110
		1: 21508
		Error: UNCLASSIFIED_PROBLEM: 3: dead (exit status 7)
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1175
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:281
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2104
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:225
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (3) 3: dead (exit status 7)
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

 | 

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*version/mixed/nodes=5.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by irfansharif**
Fixes #72029.

`using_applied_state_key` was previously used to check whether the Range
had upgraded to begin using the `RangeAppliedState` key. When set to true,
replicas receiving the state (through a `ReplicatedEvalResult`) knew to
begin using the new key. In #58088 (in 21.1) we introduced a migration
to iterate through all ranges in the system and have them start using
the `RangeAppliedState` key. In 21.2, this field was always set to true --
21.2 nodes were already using the `RangeAppliedState` key, or receiving
messages from 21.1 nodes that were also using the `RangeAppliedState` key.
In 21.2 (and in 21.1 for that matter) we didn't need to trigger the "if
set to true in an incoming message, start using the `RangeAppliedState`
key" code path.

When looking to get rid of this field in 22.1 (#70464), we observed that
there was an unintentional read of this field in 21.2 nodes (see #72029 and
\#72222); the saga is as follows:
 - Removing this field in 22.1 meant it was set as false when received at
   21.2 nodes.
 - This should've been fine! We weren't using this field to trigger any
   upgrade code paths (like it was originally intended for).
 - It turns out that in 21.2 we were using the `ReplicaState` from the
   incoming snapshot to update our in-memory replica state
 - Because the proto field was being phased out, there was now a divergence
   between the on-disk state (field set to true, from earlier 21.2
   operations) and the in-memory state (field set to false, because sent
   from a version that attempted to get rid of this field).

Removing proto fields from the replica state are not possible until we stop
using the protobuf copy of the replica state when applying a snapshot
(#72222). Once that's done, we should be able to stop sending the replica
state as part of the snapshot in the subsequent release.

Release note: None

### CODE REVIEW

**Review by erikgrinaker — APPROVED** [review]
Reviewing from my phone, so haven't checked the surrounding code, but LGTM to fix the nightly tests. Will have a closer look later.

**Review by nvb — APPROVED** [review]
:lgtm:

> Truly a migration from hell.

Truly a migration ~from hell.~

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72239#-:-MnC7T1S5yCeNyuNl8-V:b-dbw7df)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

File: pkg/kv/kvserver/replica_command.go [hunk 1]
```diff
@@ -2451,6 +2451,10 @@ func (r *Replica) sendSnapshot(
 		Term:  snap.RaftSnap.Metadata.Term,
 	}
 
+	// See comment on DeprecatedUsingAppliedStateKey for why we need to set this
+	// explicitly for snapshots going out to followers.
+	snap.State.DeprecatedUsingAppliedStateKey = true
```
Comment (tbg) [inline pkg/kv/kvserver/replica_command.go:2456]: I think if you did this in `stateloader.Load` you wouldn't have to think about this anywhere else.

**Review by tbg — APPROVED** [review]
Had a comment to, I think, improve this but let's merge anyway to get master green.

File: pkg/kv/kvserver/replica_command.go (hunk 1 above)
Comment (irfansharif) [reply pkg/kv/kvserver/replica_command.go:2456]: Yea, but I wanted this code to be close to the follower snapshot business.

**Comment by irfansharif** [comment]
TFTRs!

```
--- PASS: version/mixed/nodes=5 (247.76s)
18:03:55 test_runner.go:515: [w0] test passed
```

bors r+

**Pull request merged**